---
title: An R Markdown document converted from "/Users/mqbsxsm2/RNA-seq_Analysis/rprogramming-main/Introduction-to-ggplot2.ipynb"
output: html_document
---

<img src="images/banner_introRProg.png" align="left" />

<table style="float:right;">
    <tr>
        <td>                      
            <div style="text-align: right"><a href="https://www.research.manchester.ac.uk/portal/syed.murtuzabaker.html" target="_blank">Syed Murtuza Baker</a></div>
            <div style="text-align: right">Research Fellow</div>
            <div style="text-align: right">University of Manchester</div>
         </td>
         <td>
             <img src="images/Syed_Baker.jpg" width="50%" />
         </td>
     </tr>
</table>

# Introduction to ggplot2
****

#### About this Notebook

This notebook would introduce `ggplot2`, one of the most widely used data visualization package.

Level: <code>beginner</code> 

Duration: Approximately 2 hours to complete

<div class="alert alert-block alert-warning"><b>Learning Objectives:</b> 
<br/> At the end of this notebook you will be able to:
    
- Describe the features of the ggplot2 package
    
- Plot in R 
    
- Explain which plots require statistical transformations


</div> 

<a id="top"></a>

<b>Table of contents</b><br>

1.0 [Introduction](#intro)

2.0 [About ggplot2](#ggplot2)

3.0 [Your Turn](#yourturn)

*****

<a id="intro"></a>

## Introduction

R is one of the leading programming languages in Biological data analysis. It is used to perform data analyis, statistics, machine learning and visualisations. This course is designed as an introduction to R for participants with no previous programming experience. 


`ggplot2` is a powerful and a flexible R package, implemented by Hadley Wickham, for producing elegant graphics piece by piece (Wickham et al. 2017).


For help in using this Jupyter notebook please refer to the [Jupyter Notebook User Guide](https://online.manchester.ac.uk/bbcswebdav/orgs/I3116-ADHOC-I3HS-HUB-1/Jupyter%20Notebooks/content/index.html#/)





*****
[back to the top](#top)

### Load the libraries

We first load the `tidyverse` package. `tidyverse` package is a collection of many packages among which `ggplot2` is one. If you want, you can only load the `ggplot2` package as well. 

```{r}
library(tidyverse)
```

We additionally load the `dplyr` package. `dplyr` is called __grammar of data manipulation__ that provides a consistent set of verbs that help you to manipulate data in many different ways. `dplyr` is also loaded with `tidyverse` but we are just showing it here that you can load it seprately as well. We will use some verbs from `dplyr` package in this notebook to manipulate our data in a certain way before plotting it using `ggplot2`.

```{r}
library(dplyr)
```

Next we load the `scater` package. `scater` is a bioconductor package for doing various sets of analysis on single cell RNA-seq data. It especially focuses on quality control and visualization of single cell data. The dummy data we will be using to learn our `ggplot2` plotting functions are a `SingleCellExperiment` object that is holding single cell gene expression data from CMML (Chronic myelomonocytic leukaemia) patients and healthy controls. Let's discuss what bioconductor, single cell RNA-seq and `SingleCellExperiment` object is.

- Bioconductor: According to wikipedia,
> _Bioconductor is a free, open source and open development software project for the analysis and comprehension of genomic data generated by wet lab experiments in molecular biology. Bioconductor is based primarily on the statistical R programming language, but does contain contributions in other programming languages._


- Single cell RNA-seq: Single cell RNA-seq is a technique to measure the gene expression of each individual cell in a sample. Single cell RNA-seq helps to characterize each of the cells in a heterogeneous cell population



- `SingleCellExperiment` class: According to Bioconductor,
> The `SingleCellExperiment` class is a lightweight Bioconductor container for storing and manipulating single-cell genomics data. It's column represent cells and row represents features which can be genes, transcripts or genomic locations. Following diagram nicely shows how a single cell data is stored in `SingleCellExperiment` object

![SingleCellExperiment](images/SingleCellExperiment.png)

```{r}
library(scater)
```

We now load the `SingleCellExperiment` object. The object is stored into a file called `sce.rds`. We would load this file using `readRDS` command into the a `SingleCellExperiment` object. We are calling this object `sce`

```{r}
sce <- readRDS('/mnt/sce.rds')
sce
```

Let us explore our `sce` object a bit. First let me print the log-normalized count for two genes, 'KLF4' and 'RUNX1' across 5 cells

```{r}
logcounts(sce)[c('KLF4','RUNX1'),1:5]
```

In the above code, `logcounts()` is a built-in function that takes out the log-normalized count data from a `SingleCellExperiment` object. The column names are the cell-barcodes. Each of the cell has it's own barcode to uniquely identify them.

### Metadata of the cells
The `colData` shows all the metadata associated with the cells. Let us see the metadata for our `sce` object

```{r}
names(colData(sce))
```

There are lots of information about the cells. But we will just use the _barcode_, _Sample_, _total_counts_ and _Cluster_ metadata for these cells. So first lets extract these metadata and assign it to a new variable. We are naming the variable as `metadata_sce`. Please note `metadata_sce` will be a `data.frame` as we are copying from another `data.frame`

```{r}
metadata_sce <- colData(sce)[,c('barcode','Sample','Cluster','total_counts')]
```

```{r}
head(metadata_sce)
```

### Copying the gene expression values to a martrix

We now copy the logcount values to a matrix can will call it `norm_counts`. We will use this matrix for all our plotting. Also, please note that we will transpose the matrix so that the gene names will go into column and cell names will go into rows. This is because `ggplot2` plots the data based on what is in the column of a matrix

```{r}
norm_counts <- t(logcounts(sce))
```

Let us again see what is in `norm_counts`. We will print the first 5 cells and the two gene like before. But now the genes are in column and cells are in row

```{r}
norm_counts[1:5, c('KLF4','RUNX1')]
```

### Data preperation

We will now combine the cell metada and the count data of three genes into one dataframe. This would make our plotting easier

```{r}
d_exp <- as_tibble(metadata_sce)
d_exp <- cbind(d_exp, norm_counts[,c('KLF4','RUNX1','EGR1')])
```

Let's look how `d_exp` dataframe looks like

```{r}
head(d_exp)
```

We will now use this data for our ggplot2 plotting.


*****
[back to the top](#top)

<a id="ggplot2"></a>

## Basics of `ggplot2`


The `gg` in `ggplot2` means __Grammar of Graphics__, a graphic concept which describes plots by using a “grammar”. According to the ggplot2 concept, a plot can be divided into different fundamental parts:


>Plot = Data + Aesthetics + Geometry

* Data: a data frame with the data in the columns which we want to plot
- Aesthetics: used to indicate the x and y variables. It can be also used to control the color, the size and the shape of points, etc…..
- Geometry: corresponds to the type of graphics (scatter plot, histogram, box plot, line plot, ….). Additional layers for customization can be done with — `title`, `labels`, `axis`, etc.

### First Plotting

The main function in the ggplot2 package is `ggplot()`, which can be used to initialize the plotting system with data and `x/y` variables.

`ggplot()` creates a coordinate system to which you can add layers to. The first argument of `ggplot()` is the dataset to use in the graph. So `ggplot(data = d_exp`) creates an empty graph (you just see a gray frame).

For example, the following R code takes the _KLF4_ and _RUNX1_ data set to initialize the ggplot and then a layer (`geom_point()`) is added onto the ggplot to create a scatter plot of `x = KLF4` by `y = RUNX1`:

```
Data= d_exp
Aesthetic=: aes(x=KLF4, y=RUNX1)
Geometry= geom_point()
```

```{r}
ggplot(d_exp)
```

Running the above code just gives us an empty frame. This is because we have just told ggplot which data we will use for plotting but have not yet plotted anything. We just filled the _Data_ part of ggplot.

Now we set the Aesthetic part of ggplot

```{r}
ggplot(d_exp, aes(x=KLF4, y=RUNX1))
```

Now we have told ggplot that we will use the gene expression value of KLF4 and RUNX1 to plot our data. But we have not told it whether the plot will be a scatter plot or a line plot or something else. So ggplot just sets the x-axis and y-axis with the gene names but have not plotted anything else. So far we have set the _Data_ and the _Aesthetic_ part of ggplot. Now we will set the final part which is the _Geometry_ part.

```{r}
ggplot(d_exp, aes(x=KLF4, y=RUNX1)) + geom_point()
```

So, now got our first complete ggplot. The x-axis shows the expression of KLF4 and the y-axis shows the expression of RUNX1. Some cells are not expressing KLF4 or RUNX1 at all, so the gene expression values for those cells are zero. That is why we see dots around the zero line.

```{r}
ggplot(d_exp, aes(x=KLF4, y=RUNX1)) + geom_point(size = 1.2, color = "steelblue", shape = 21)
```

You can see that the dots are now hollow. This is a different shape we have chosen here using the parameter `shape=21`. Also we changed the color of the dots to `steelblue`. There are lots of colours you can choose for your plots. A complete list of colours can be found in [R Color Palette](https://www.nceas.ucsb.edu/sites/default/files/2020-04/colorPaletteCheatsheet.pdf).

We can also store the plot in a variable, say `p`, so that we can print it later when we need it

Let us now try to colour the dots based on samples.

```{r}
# Control points color by groups
ggplot(d_exp, aes(x=KLF4, y=RUNX1))+
  geom_point(aes(color = Sample))
```

In the plot above we coloured the dots based on samples. So each dots are coloured based on the sample it is coming from. You can see the color legend on the right of the figure. 

Now what if you do not like the default colours R is making, what if you want to change it to colours you like. You can change the default colours using `scale_color_manual`. Within `scale_color_manual` you can put the colour names from the color palette cheetsheet I mentioned earlier or you can use the HTML colour codes.

So what are the HTML colour code?

Each of the colour is represnted in the format __RGB__ where R is for red, G is for Green and B is for Blue. Each of the colour will take two hexa-decimal values (0 - 9 and then A - F). Each of the value creates a shade of the colour. For example, in the color red, the color code is #FF0000, which is '255' red, '0' green, and '0' blue. You need to start with a "#" sign to indicate that you are putting a colour code and not just a simple string.

We had 10 samples so we choose 10 colours below. Change the colour code to play with different values

```{r}

# Change the default color manually.
# Use the scale_color_manual() function
p <- ggplot(d_exp, aes(x=KLF4, y=RUNX1))+
  geom_point(aes(color = Sample))+
  scale_color_manual(values = c("#00AFBB", "#E7B800", "#FC4E07", "#A2AFBB", "#17B8AB", 
                                "#3F4E77", "#FCFA27", "#BFAFFB", "#69B89B", "#7F4E97"))
print(p)
```

*****
[back to the top](#top)

### GGPlot theme

So far we have worked with only the `ggplot2` default theme that produces a grid with gray background. In this section we will learn how to chnage those themes.

As said before the default theme of ggplots is `theme_gray(`) (or `theme_grey()`), which is theme with grey background and white grid lines. More themes are available for professional presentations or publications. These include: `theme_bw(`), `theme_classic()` and `theme_minimal()`.

To change the theme of a given ggplot (p), we will use: `p + theme_classic()`.

```{r}
p <- ggplot(d_exp, aes(x=KLF4, y=RUNX1))+
  geom_point(aes(color = Sample))+
  scale_color_manual(values = c("#00AFBB", "#E7B800", "#FC4E07", "#A2AFBB", "#17B8AB", "#3F4E77", "#FCFA27", "#BFAFFB", "#69B89B", "#7F4E97"))
p <- p + theme_classic()
print(p)
```

You can see that the background has changed from Gray to white and the grid lines have disappear. Now try with `theme_bw()` to see what other changes it makes. To know the details of the themes you can go to [ggplot complete theme](https://ggplot2.tidyverse.org/reference/ggtheme.html)

## Adding layers to ggplot object

Now, so far we have just drawn the scater plot using `geom_point()`. One strenght of ggplot is that we can add multiple layers on top of each other to visualize other information.

For e.g. now if we want to see how a regression line would fit to the gene expression data of KLF4 and RUNX1 we can calculate a regression line and add another layer to show the line on top of the plot we already prepared.
> regression line is a linear approach for modelling the relationship between a response and explanatory variables 

We would add these layers using other `geom`. So the plot using a regression line would look like

```{r}
d_exp$pred.SC <- predict(lm(RUNX1 ~ KLF4, data = d_exp))

ggplot(d_exp, aes(x = KLF4, y = RUNX1)) + 
  geom_point(aes(color = Sample)) +
  geom_line(aes(y = pred.SC)) +
  theme_classic()
```

The `predict` is a generic function for predicting the line that fits the RUNX1 and KLF4 gene expression. We then take this fitted line and draw it using the geom `geom_line`. In this dataset, the line is not a great fit because of the zero expression by some cells.

### Setting Title, xlab & ylab

So far we have seen how to use ggplot2 and draw some basic plots. The label for the x-axis & y-axis was set by default by `ggplot`. Now what if we want to change the name of our axis to something else, we can do that using `xlab()` and `ylab` function. Also if we want to set the title of the plot we can do it using `ggtitle`

```{r}
d_exp$pred.SC <- predict(lm(RUNX1 ~ KLF4, data = d_exp))

ggplot(d_exp, aes(x = KLF4, y = RUNX1)) + 
  geom_point(aes(color = Sample)) +
  geom_line(aes(y = pred.SC)) +
  xlab('our KLF4 expression') +
  ylab('our RUNX1 expression') +
  ggtitle('nice plotting of gene expression values')+
  theme_classic()
```

### Histogram

We will now also see some other useful geoms. We will start with `geom_histogram()`. `geom_histogram()` draws a histogram on the data. 

```{r}
ggplot(d_exp, aes(x=total_counts)) + geom_histogram() + theme_classic() 
```

So, we just plotted our first histogram using ggplot2. But it is showing an error. The bins it set by default is not happy with it and asking us to change it. We try again after setting the binwidth

```{r}
ggplot(d_exp, aes(x=total_counts)) + geom_histogram(bins=30) + theme_classic() 
```

So, even not chaning the bin, but just setting this parameter as `geom_histogram(bins=30)` clears the warning. This also tells us sometimes making small changes might solve issues.

### Density plot

A density plot shows the distribution of a numeric variable. It uses a __kernel density estimate__ to show the probability density function of the variable. Here we will estimate the desnity of total_counts which is the sequencing depth for each of the sample.

```{r}
df <- data.frame(x=log10(sce$total_counts+1), Sample = sce$Sample)
ggplot(df,
       aes(x = x, fill = as.factor(Sample))) + 
       geom_density(alpha = 0.5) +
       labs(x = expression('log'[10]*'(Library Size)'), title = "Total reads density", fill = "Sample") + 
       theme_classic(base_size = 14) + # Setting the base size text for plots
       scale_fill_manual(values = c("#00AFBB", "#E7B800", "#FC4E07", "#A2AFBB", "#17B8AB", "#3F4E77", "#FCFA27", "#BFAFFB", "#69B89B", "#7F4E97")) # Need to set fill manual
```

Majority of the samples have similar sequencing depth except sample BC572. We know this sample has technical issues during sample collection, so it is no surprise.

But the density is overlapping with each other and it is hard to clearly view the density of each sample. This is where __Facet__ becomes useful.

### Facet

Facet split a plot into matrix of panels. In the example below we will split the density of each sample into multiple facets to clearly see the density of their library size.

```{r}
df <- data.frame(x=log10(sce$total_counts+1), Sample = sce$Sample)
ggplot(df,
       aes(x = x, fill = as.factor(Sample))) + 
       geom_density(alpha = 0.5) +
       labs(x = expression('log'[10]*'(Library Size)'), title = "Total reads density", fill = "Sample") + 
       theme_classic(base_size = 14) + # Setting the base size text for plots
       scale_fill_manual(values = c("#00AFBB", "#E7B800", "#FC4E07", "#A2AFBB", "#17B8AB", "#3F4E77", "#FCFA27", "#BFAFFB", "#69B89B", "#7F4E97")) +  # Need to set fill manual 
  facet_wrap(~Sample)
```

*****
[back to the top](#top)

## Statistical Transformations

Some plot types (such as scatterplots) do not require transformations, each point is plotted at x and y coordinates equal to the original value. But some other plots, such as boxplots, histograms, prediction lines etc. require statistical transformations. In this section we will learn about these statistical transformations.

We start with `geom_smooth()`. This function tries to plot a non-linear curve (not a stright line) to our gene expression values. By default, `geom_smooth()` would use a genralized additive model. It will basically try to fit a non-linear curve to the expression values.

```{r}
#d_exp$pred.SC <- predict(lm(RUNX1 ~ KLF4, data = d_exp))

ggplot(d_exp, aes(x = KLF4, y = RUNX1)) + 
  geom_point(aes(color = Sample)) +
  geom_smooth() +
  theme_classic()
```

But now what if we want to fit a linear curve? For that we choose the parameter "lm" or linear model.

```{r}
ggplot(d_exp, aes(x = KLF4, y = RUNX1)) + 
  geom_point(aes(color = Sample)) +
  geom_smooth(method = "lm") +
  theme_classic()
```

Prevously we did the same thing when we were trying to predict the linear line going through the points. We used the `predict` function over there and then used `geom_line` to draw the line. But for `geom_smooth()` we can actually just pass the method and the geom would take care of the rest. 

# Resources

- One of the best place to learn about different ggplot functions are to look at RStudio's [ggplot cheetsheet](https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-visualization.pdf)
- Also the course at r-statistics.co is very helpful, [The Complete ggplot2 Tutorial - Part1](http://r-statistics.co/Complete-Ggplot2-Tutorial-Part1-With-R-Code.html)

<a id="yourturn"></a>
## Your Turn

First let us look at the mpg dataset

```{r}
head(mpg)
```

<div class="alert alert-block alert-info">
    <b>Task 1</b></br>
<p>How would you plot the relationship between cty, the average city mileage, and hwy, the average highway mileage? How would you describe this relationship?</p>
</div>

```{r}
ggplot(mpg, aes(cty, hwy)) + 
  geom_point() +
  theme_classic()
```

<div class="alert alert-block alert-info">
    <b>Task 2</b> </br>
<p>How would you draw a smooth line? Also, how would you colour the transmissions differently?</p>

```{r}
ggplot(mpg, aes(cty, hwy)) + 
  geom_point(aes(colour = drv)) +
  geom_smooth(method = "lm") +
  theme_classic()
```

<div class="alert alert-block alert-info">
    <b>Task 3</b> </br>
    <p>Show the distribution of scores for hwy, highway miles per gallon</p></div>

```{r}
ggplot(mpg, aes(hwy)) + 
  geom_histogram()
```

<div class="alert alert-block alert-info">
    <b>Task 4</b> </br>
    <p>Produce a similar plot as shown below</p>
</div>
<img src="images/task4-ggplot.png" align="left" width="80%" />

```{r}
ggplot(mpg, aes(x=displ, y=cty, colour=drv, size=cyl)) +
  geom_point() + theme_classic()
```

*****
[back to the top](#top)

### Notebook details
<br>
<i>Notebook created by <strong>Syed Murtuza Baker</strong>. Other contributors include Fran Hooley... 

Publish date: May 2021<br>
Review date: May 2022</i>

Please give your feedback using the button below:

****

## Your Notes:

